synapse/rest/client/v2_alpha/account.py

```diff
@@ -690,6 +703,115 @@ def _add_threepid_to_account(self, user_id, validation_session):
         )
 
 
+class ThreepidAddRestServlet(RestServlet):
+    PATTERNS = client_patterns("/account/3pid/add$", releases=(), unstable=True)
+
+    def __init__(self, hs):
+        super(ThreepidAddRestServlet, self).__init__()
+        self.hs = hs
+        self.identity_handler = hs.get_handlers().identity_handler
+        self.auth = hs.get_auth()
+        self.auth_handler = hs.get_auth_handler()
+        self.store = self.hs.get_datastore()
+
+    @defer.inlineCallbacks
+    def on_POST(self, request):
+        body = parse_json_object_from_request(request)
+
+        assert_params_in_dict(body, ["client_secret", "sid"])
+        client_secret = body["client_secret"]
+        sid = body["sid"]
+        threepid_creds = {"client_secret": client_secret, "sid": sid}
+
+        requester = yield self.auth.get_user_by_req(request)
+        user_id = requester.user.to_string()
+
+        # We don't actually know which medium this 3PID is. Thus we first assume it's email,
```

Comment:
hum, I'm not really happy with the duplication with /account/3pid, even if it is only temporary. Even temporary things have a habit of sticking around for long enough to be a pain. Can you factor out the duplication? The easiest way might be to move everything below this point out to the identity handler.

Reply:
Yes, sorry. It was temporary in the sense that I was waiting for #6042 to land.

Reply:
Have been refactored out into a new method in IdentityHandler.